### PR TITLE
add congestion info fields to chunk header

### PR DESCRIPTION
**Pull request by jakmeier**
This is a part of NEP-539.

This PR alone only adds the fields, without actually making use of them. But it adds all the necessary data for the congestion control strategy to be able to function. This makes this PR by far the largest.

Splitting the feature into multiple PRs is supposed to make reviewing it easier.

### CODE REVIEW

**Comment by wacban** [comment]
Here is how the data flows and propagates the congestion info in the system. Please keep in mind that there are two types of congestion info - that of a single shard (CongestionInfo) and that of all shards (HashMap<ShardId, CongestionInfo>). 

The flow of data for putting congestion info into ChunkExtra, from the bottom (runtime) to the top (client):
* `Runtime::apply` - the congestion info is placed into ApplyResult. Currently it's only the default, in future PRs it will be calculated based on the congestion info from the previous block and the receipts added and removed from the queues and buffers. 
* `NightshadeRuntime::process_state_update` - the congestion info is passed from ApplyResult to ApplyChunkResult
* `NightshadeRuntime::apply_chunk` - the congestion info is passed within ApplyChunkResult
* the execution now forks in two
  * regular processing path
    * `apply_new_chunk` and `apply_old_chunk` put the congestion info into NewChunkResult or OldChunkResult
    * `process_shard_update` puts the congestion info into NewChunk or OldChunk
    * `get_update_shard_job` schedules the above and puts in the db to be picked up in postprocess_block
    * `postprocess_block` -> `apply_chunk_postprocessing` - > `process_apply_chunk_result`
    * `process_apply_chunk_result` stores the congestion info in ChunkExtra
  * state sync processing path
    * TBD

The flow of data for passing all shards congestion info from the prev block to the chunk application:
* the first few steps are the same as the last few steps in the the previous flow in reverse
* `NightshadeRuntime::process_state_update` unpacks ApplyChunkBlockContext and extracts congestion info. The congestion info is then stored in the `ApplyState`
* `Runtime::apply` - upacks the ApplyState and can use congestion info. 

The flow of data for passing all shards congestion info from the prev block to the chunk production: 
* `Client::produce_chunk` gets the previous block which contains ChunkHeaders with the congestion info. 
* the prev block is passed as context to `prepare_transactions` where congestion info will be used to filter transactions

The flow of data for passing congestion info from the ChunkExtra of the previous chunk to the ChunkHeader of the next chunk
* `Client::produce_chunk` gets the previous ChunkExtra with the congestion info. 
* the congestion info from the previous chunk extra is passed to `create_encoded_shard_chunk` to be included in the current chunk header 

TODO:
  * stateless validation 
  * resharding 
  * state sync

File: chain/client/src/stateless_validation/chunk_validator/mod.rs
Comment (jakmeier) [inline chain/client/src/stateless_validation/chunk_validator/mod.rs:438]: Huh? Did we reach a threshold with the new fields in chunk extra, or how did this trigger suddenly?

File: chain/client/src/stateless_validation/chunk_validator/orphan_witness_handling.rs
Comment (jakmeier) [inline chain/client/src/stateless_validation/chunk_validator/orphan_witness_handling.rs:121]: Has this something to do with the new congestion control fields. or is it just an on-the-fly observation?

File: chain/client/src/test_utils/client.rs
Comment (jakmeier) [inline chain/client/src/test_utils/client.rs:210]: again, please qualify before merging
```suggestion
            // TODO(congestion_control): compute if not available
```

File: core/primitives/src/congestion_info.rs
Comment (jakmeier) [inline core/primitives/src/congestion_info.rs:11]: :heart:

File: core/primitives/src/sharding.rs
Comment (jakmeier) [inline core/primitives/src/sharding.rs:459]: Looks correct, but I gotta say, this version handling is getting is a bit insane :S
But not your fault ;)

Nit: Personally, I find it easier to read when cases with identical code are summarized. But feel free to keep at is if you like them separate.

```suggestion
            ShardChunkHeader::V3(header) => match header.inner {
                ShardChunkHeaderInner::V1(_) | ShardChunkHeaderInner::V2(_) => {
                    version >= BLOCK_HEADER_V3_VERSION && version < CONGESTION_CONTROL_VERSION
                }
                ShardChunkHeaderInner::V3(_) => version >= CONGESTION_CONTROL_VERSION,
            },
```

**Review by jakmeier — COMMENTED** [review]
Looking already much better than what I created! I think it's good timing to get it merged in this state and iterate further with smaller PRs, it will make many things easier.

I left several comments but they are only minor details to quickly address before merging.
I would approve if I could, but since I originally submitted the PR, it won't let me...
@wacban can you self-approve as a co-author?

**Review by wacban — APPROVED** [review]
self approve :)

**Pull request merged**
